Add Catalan numbers

dlesnoff opened the pull request:
I have implemented all formulas to compute Catalan Numbers and have given a compile-time table computation.
I have also implemented an iterator version.
It might be too much, please let me know what you want to keep.

For comparison, I link below the existing implementations:
Math version: https://the-algorithms.com/algorithm/catalan-number
Dynamic programming: https://the-algorithms.com/algorithm/catalan-numbers
On Rosetta Code: https://rosettacode.org/wiki/Catalan_numbers

Panquesito7 commented on dynamic_programming/catalan_numbers.nim:
```diff
@@ -0,0 +1,105 @@
+        inc i
+    
+    test "Uses compile-time table":
+      check catalan_number(36) == Positive(3116285494907301262)
```
Missing an endline. Without this endline, there could be issues with some IDEs and configurations.
```suggestion
      check catalan_number(36) == Positive(3116285494907301262)
      
```

dlesnoff replied:
Done.

Panquesito7 reviewed (approved):
I'm not proficient with Nim, but this looks good. Thanks. 🚀
Will wait until other people check it out, if not, I'll merge it in the next few days.

ZoomRmc commented on dynamic_programming/catalan_numbers.nim:
```diff
@@ -0,0 +1,105 @@
+  ## Returns the index-th Catalan number recursively.
+  ## As Nim does not make automatic memoization, this function is not
+  ## efficient.
+  if index == 0:
+    return 1
+  for i in 0 ..< index:
+    result += catalan_numbers_recursive(i) *
+     catalan_numbers_recursive(index - i - 1)
```
```suggestion
  if index < 2:
    return 1
  for i in 0 ..< index:
    result += catalanNumbersRecursive(i) *
     catalanNumbersRecursive(index - i - 1)
```

ZoomRmc commented on dynamic_programming/catalan_numbers.nim:
```diff
@@ -0,0 +1,105 @@
+     catalan_numbers_recursive(index - i - 1)
+
+func catalanNumbersRecursive2(index: Natural): Positive =
+  if index == 0:
```
```suggestion
  if index < 2:
```

ZoomRmc commented on dynamic_programming/catalan_numbers.nim:
```diff
@@ -0,0 +1,105 @@
+  else:
+    2 * (2 * index - 1) * catalanNumbersRecursive2(index - 1) div (index + 1)
+
+func catalanNumbers(index_limit: Natural): seq[Positive] =
+  ## Returns all Catalan numbers up to the index-th number iteratively.
+  var catalan_numbers = newSeq[Positive](index_limit + 1)
+  catalan_numbers[0] = 1
+  for i in 1 .. index_limit:
+    for j in 0 ..< i:
+      catalan_numbers[i] += catalan_numbers[j] * catalan_numbers[i - j - 1]
+  catalan_numbers
```
I'm not a fan of `result`s but since we're using them anyway... Not reusing an identifier is a bonus in this case.
```suggestion
func catalanNumbers(indexLimit: Natural): seq[Positive] {.noinit.} =
  ## Returns all Catalan numbers up to the index-th number iteratively.
  result = newSeq[Positive](indexLimit + 1)
  result[0] = 1
  for i in 1 .. indexLimit:
    for j in 0 ..< i:
      result[i] += result[j] * result[i - j - 1]
```

ZoomRmc commented on dynamic_programming/catalan_numbers.nim:
```diff
@@ -0,0 +1,105 @@
+
+iterator catalanNumbersIt(index: Natural): Positive =
+  ## Iterates over all Catalan numbers up to the index-th number.
+  var catalan_numbers_lst = newSeq[Positive](index + 1)
+  catalan_numbers_lst[0] = 1
+  yield 1
+  for i in 1 .. index:
+    for j in 0 ..< i:
+      catalan_numbers_lst[i] += catalan_numbers_lst[j] * catalan_numbers_lst[i - j - 1]
+    yield catalan_numbers_lst[i]
```
Multiple yields in regular Nim iterators should be avoided as they can lead to considerable code duplication. Here it's doable and is even shorter.

```suggestion
  var catalanNumbers = newSeq[Positive](index + 1)
  catalanNumbers[0] = 1
  for i in 0 .. index:
    for j in 0 ..< i:
      catalanNumbers[i] += catalanNumbers[j] * catalanNumbers[i - j - 1]
    yield catalanNumbers[i]
```

Panquesito7 commented:
Thanks for your review, @ZoomRmc! 🚀

ZoomRmc commented:
Thanks! Keep on doing the good work, @dlesnoff!

dlesnoff commented:
I extensively tested the code.
I added limits for the different methods.
To compute directly a number for a specific index, the other formulas are probably better, but they overflow due to an intermediate computation. I added these overflow limits as consts.

I fix a defect in the use of the compile time table.

I made indexes consistent with the comment, and stop the computations before the input parameter.
